cirq-core/cirq/experiments/readout_confusion_matrix_test.py

```diff
@@ -0,0 +1,137 @@
+
+import numpy as np
+import cirq
+import pytest
+
+from cirq.experiments.single_qubit_readout_calibration_test import NoisySingleQubitReadoutSampler
+
+
+def get_expected_cm(num_qubits: int, p0: float, p1: float):
+    expected_cm = np.zeros((2 ** num_qubits,) * 2)
+    for i in range(2 ** num_qubits):
+        for j in range(2 ** num_qubits):
+            p = 1.0
+            for k in range(num_qubits):
+                b0 = (i >> k) & 1
+                b1 = (j >> k) & 1
+                if b0 == 0:
+                    p *= p0 * b1 + (1 - p0) * (1 - b1)
+                else:
+                    p *= p1 * (1 - b1) + (1 - p1) * b1
+            expected_cm[i][j] = p
+    return expected_cm
+
+
+@pytest.mark.parametrize('p0, p1', [(0, 0), (0.2, 0.4), (0.5, 0.5), (0.6, 0.3), (1.0, 1.0)])
+def test_measure_confusion_matrix_with_noise(p0, p1):
+    sampler = NoisySingleQubitReadoutSampler(p0, p1, seed=1234)
+    num_qubits = 4
+    qubits = cirq.LineQubit.range(num_qubits)
+    expected_cm = get_expected_cm(num_qubits, p0, p1)
+    qubits_small = qubits[:2]
+    expected_cm_small = get_expected_cm(2, p0, p1)
+    repetitions = 12_000
+    # Build entire confusion matrix by running 2 ** 4 = 16 circuits.
+    readout_cm = cirq.measure_confusion_matrix(sampler, qubits, repetitions=repetitions)
+    assert readout_cm.repetitions == repetitions
+    for q, expected in zip([None, qubits_small], [expected_cm, expected_cm_small]):
+        np.testing.assert_allclose(readout_cm.confusion_matrix(q), expected, atol=1e-2)
+        np.testing.assert_allclose(
+            readout_cm.confusion_matrix(q) @ readout_cm.correction_matrix(q),
+            np.eye(expected.shape[0]),
+            atol=1e-2,
+        )
+
+    # Build a tensored confusion matrix using smaller single qubit confusion matrices.
+    # This works because the error is uncorrelated and requires only 4 * 2 = 8 circuits.
+    readout_cm = cirq.measure_confusion_matrix(
+        sampler, [[q] for q in qubits], repetitions=repetitions
+    )
+    assert readout_cm.repetitions == repetitions
+    for q, expected in zip([None, qubits_small], [expected_cm, expected_cm_small]):
+        np.testing.assert_allclose(readout_cm.confusion_matrix(q), expected, atol=1e-2)
+        np.testing.assert_allclose(
+            readout_cm.confusion_matrix(q) @ readout_cm.correction_matrix(q),
+            np.eye(expected.shape[0]),
+            atol=1e-2,
+        )
+
+    # Apply corrections to sampled probabilities using readout_cm.
+    qs = qubits_small
+    circuit = cirq.Circuit(cirq.H.on_each(*qs), cirq.measure(*qs))
+    reps = 100_000
+    sampled_result = cirq.get_state_histogram(sampler.run(circuit, repetitions=reps)) / reps
+    expected_result = [1 / 4] * 4
+
+    def l2norm(result: np.ndarray):
+        return np.sum((expected_result - result) ** 2)
+
+    corrected_result = readout_cm.apply(sampled_result, qs)
+    assert l2norm(corrected_result) <= l2norm(sampled_result)
```

Comment:
Isn't the value of l2norm(corrected_result) bounded by the repetitions and we can get arbitrary precision here? Maybe it would be worth checking for a tighter bound?

Reply:
There's sampling error in the confusion matrix computation + sampling error in the `sampled_result`. Increasing repetitions of confusion matrix computation has a larger overhead and make the tests very slow very quickly. Increasing repetitions for sampling the final circuit does not make the result very close the expected result. Hence I found this assertion to be a good tradeoff b/w speed and correctness.